JS Learning Area Objects module updates (#11373)
mdn/content: files/en-us/learn/javascript/objects/object_prototypes/index.md

      <th scope="row">Objective:</th>
      <td>
        To understand JavaScript object prototypes, how prototype chains work,
        and how to set the prototype of an object.
      </td>
    </tr>
  </tbody>
</table>

## The prototype chain

In the browser's console, try creating an object literal:

```js
const myObject = {
  city: 'Madrid',
  greet() {
    console.log(`Greetings from ${this.city}`);
  }
}

myObject.greet(); // Greetings from Madrid
```

This is an object with one data property, `city`, and one method, `greet()`. If you type the object's name and a period into the console, like `myObject.`, then the console will show you a list of all the properties available to this object. You'll see that as well as `city` and `greet`, there are lots of other properties!

> ```

Reviewer: When I type in myObject I do see lots of other items, but they are grouped under this `[[prototype]]` thingy which I assume represents `__proto__`. Is this mismatch in what we see something that needs to be explained?

![image](https://user-images.githubusercontent.com/5368500/147722792-377fc31e-bae7-476b-a034-9f3621c4bffd.png)

Author: In the description:

> If you type the object's name and a period into the console, like `myObject.`, then the console will show you a list of all the properties available to this object.

...the period is important. If you just type `myObject` and return, then yes, you get the console's representation of the object. If you type `myObject.` without a return, you get the autocomplete popup, which is what I was trying to show.

I guess this is tricky! Maybe it would be better to do something like:

> This is an object with one data property, `city`, and one method, `greet()`.
>
> Now try typing this into the console:
>
> ```js
> console.log(myObject.hasOwnProperty());
> console.log(myObject.toString());                  
> ```
> You should see something like this:
>
> ```
> false
> [object Object]
> ```
> What are these extra functions, and where do they come from?

It's a bit of a shame not to show all the extra properties, but this is probably easier to follow. What do you think?

Reviewer: No, I think you've done it just right - I just missed the period because I assumed I knew how this works. I have proposed using style to make this more obvious https://github.com/mdn/content/pull/11373/files#r777753158

Reviewer: I'll mark this as resolved, up to you if you like the proposed fix.

__defineGetter__
__defineSetter__
__lookupGetter__
__lookupSetter__
__proto__
city
constructor
greet
hasOwnProperty
isPrototypeOf
propertyIsEnumerable
toLocaleString
toString
toValueOf
```

Try accessing one of them:

```js
myObject.hasOwnProperty('city'); // true
```

It works (even if it's not obvious yet what `hasOwnProperty()` does).

What are these extra properties, and where do they come from?

Every object in JavaScript has a built-in property, which is called its **prototype**. The prototype is itself an object, so the prototype will have its own prototype, making what's called a **prototype chain**. The chain ends when we reach a prototype that has `null` for its own prototype.

> > **Note:** The property of an object that points to its prototype is **not** called `prototype`. Its name is not standard, but in practice all browsers use {{jsxref("Object/__proto__", "__proto__")}}. The standard way to access an object's prototype is the {{jsxref("Object/getPrototypeOf", "Object.getPrototypeOf()")}} method.

Reviewer: So to be clear:
- An object's prototype property is pointed called `/__proto__`
- A constructor function property named `prototype` will be set to the prototype of the newly created object.
- So if you construct an object from a constructor, under the hood there will be an assignment that looks like `newobject.__proto__=constructorfunction.prototype`

Is that right?

Author: > So to be clear:
> 
>     * An object's prototype property is pointed called `/__proto__`
> 
>     * A constructor function property named `prototype` will be set to the prototype of the newly created object.
> 
>     * So if you construct an object from a constructor, under the hood there will be an assignment that looks like `newobject.__proto__=constructorfunction.prototype`
> 
> 
> Is that right?
> 

100%.

> I think I found it jarring that we mentioned that the object prototype property is not named prototype, and then in the constructor we have a property named prototype that assigns the prototype for the created object.

The fact that the constructor has a property `prototype` is exactly why we had that note - the name of the property that references the prototype is `__proto__` not `prototype`.

> But then it occurs to me that the constructor.prototype is NOT the constructors prototype - that would be named `__proto__`.

Yes, exactly.

> I think I might be thick.

No, this stuff is very confusing.

Reviewer: Cool. I think perhaps we can help a bit by attempting to reiterate the difference between `prototype` and `__proto__` again - I have suggested this in https://github.com/mdn/content/pull/11373/files#r777761617 - for your consideration.

  }
}

const carl = Object.create(personPrototype);
carl.greet();  // hello!
```

Here we create an object `personPrototype`, which has a `greet()` method. We then use `Object.create()` to create a new object with `personPrototype` as its prototype. Now we can call `greet()` on the new object, and the prototype provides its implementation.

### Using a constructor

In JavaScript, all functions have a property named `prototype`. When you call a function as a constructor, the prototype of the object the constructor creates is set to the constructor's `prototype`.

So if we set the `prototype` of a constructor, we can ensure that all objects created with that constructor are given that prototype:

```js
const personPrototype = {
  greet() {
    console.log(`hello, my name is ${this.name}!`);
  }
}

function Person(name) {
  this.name = name;
}

Person.prototype = personPrototype;
Person.prototype.constructor = Person;
```

Here we create:
* an object `personPrototype`, which has a `greet()` method
* a `Person()` constructor function which initializes the name of the person to create.

We then set the `Person` function's `prototype` property to point to `personPrototype`.

> The last line `Person.prototype.constructor = Person;` needed because the prototype has a `constructor` property that points to the constructor function that was used to create the object. Since we have replaced the `prototype` property, we need to set `prototype.constructor` back to its original value.

Reviewer: ```suggestion
The last line (`Person.prototype.constructor = Person;`) is needed because the prototype has a `constructor` property that points to the constructor function that was used to create the object. Since we have replaced the `prototype` property, we need to set `prototype.constructor` back to its original value.
```

This bit confuses me a little:

> .. because the prototype has a `constructor` property that points to the constructor function that was used to create the object. 

So the prototype is `personPrototype` - but it is an object literal. So what constructor was used to create it? We didn't use `new` to create it and we haven't introduced that it might have one somehow under the hood. Or am I missing something? I think I get the point, but it feels a bit "inception"

Author: It is really confusing. I don't like this bit at all, because I want to treat it as an aside (which it is, really) but it's complicated enough to get snagged on.

AFAICT it's about the instance that's created, `reuben`, and what the result is of accessing its `constructor` property. For instance, try not setting `constructor` as we do above:

```js
const personPrototype = {
  greet() {
    console.log(`hello, my name is ${this.name}!`);
  }
}

function Person(name) {
  this.name = name;
}

Person.prototype = personPrototype;

const reuben = new Person('Reuben');

console.log(reuben.constructor);
```

This will give us `function Object()` (because the constructor for the `personPrototype` instance was `Object` (because we created it as an object literal)). This is obviously wrong - the constructor for `reuben` should be  `Person()`, because that's what we used to create the `reuben` instance.

I want to say here "don't look too deeply into this at this point, this is just a thing you have to do" but I never like saying this in docs. There is more detail here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/constructor#changing_the_constructor_of_a_function  but honestly I didn't find it very helpful.

After this code, objects created using `Person()` will get `personPrototype` as their prototype.

```js
const reuben = new Person('Reuben');
reuben.greet(); // hello, my name is Reuben!
```

This also explains why we said earlier that the prototype of `myDate` is called `Date.prototype`: it's the `prototype` property of the `Date` constructor.

> ### Own properties

Reviewer: FYI hasOwnProperty is supposed to be replaced eventually by [Object.hasOwn()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn). 

Might we worth mentioning here as it is better. Might not as it is not widely implemented. Perhaps a note?

Author: I wonder if I should also change the usage of `hasOwnProperty` earlier on?

Well, for now I'll just add a note down here.

Author: I just pushed https://github.com/mdn/content/pull/11373/commits/ac1ca1aca46c97249406af6d7108c624f78f0670 to talk about `hasOwn` here and add a note about `hasOwnProperty` (giving primacy to the new version, since it seems to be supported in most browsers).

Reviewer: Works for me.

      <th scope="row">Objective:</th>
      <td>
        To understand JavaScript object prototypes, how prototype chains work,
        and how to set the prototype of an object.
      </td>
    </tr>
  </tbody>
</table>

## The prototype chain

In the browser's console, try creating an object literal:

```js
const myObject = {
  city: 'Madrid',
  greet() {
    console.log(`Greetings from ${this.city}`);
  }
}

myObject.greet(); // Greetings from Madrid
```

> This is an object with one data property, `city`, and one method, `greet()`. If you type the object's name and a period into the console, like `myObject.`, then the console will show you a list of all the properties available to this object. You'll see that as well as `city` and `greet`, there are lots of other properties!

Reviewer: Perhaps - following on from https://github.com/mdn/content/pull/11373/files#r776568326  - 

```suggestion
This is an object with one data property, `city`, and one method, `greet()`. If you type the object's name _followed by a period_ into the console, like `myObject.`, then the console will pop up a list of all the properties available to this object. You'll see that as well as `city` and `greet`, there are lots of other properties!
```

};

console.log(myDate.getYear()); // 'something else!'
```

This should be predictable, given the description of the prototype chain. When we call `getYear()` the browser first looks in `myDate` for a property with that name, and only checks the prototype if `myDate` does not define it. So when we add `getYear()` to `myDate`, then the version in `myDate` is called.

This is called "shadowing" the property.

## Setting a prototype

There are various ways of setting an object's prototype in JavaScript, and here we'll describe two: `Object.create()` and constructors.

### Using Object.create

The `Object.create()` method creates a new object and allows you to specify an object that will be used as the new object's prototype.

Here's an example:

```js
const personPrototype = {
  greet() {
    console.log('hello!');
  }
}

const carl = Object.create(personPrototype);
carl.greet();  // hello!
```

Here we create an object `personPrototype`, which has a `greet()` method. We then use `Object.create()` to create a new object with `personPrototype` as its prototype. Now we can call `greet()` on the new object, and the prototype provides its implementation.

### Using a constructor

> In JavaScript, all functions have a property named `prototype`. When you call a function as a constructor, the prototype of the object the constructor creates is set to the constructor's `prototype`.

Reviewer: This is just a suggestion, perhaps worth re-iterating here 
```suggestion
In JavaScript, all functions have a property named `prototype`. When you call a function as a constructor, this property is set as the prototype of the newly constructed object (by convention, in the property named `__proto__`).
```

We might also consider a note after:

> **Note** The terminology is confusing.  The `prototype` property is the prototype that gets assigned to objects constructed from the function. The prototype of the function itself is stored in its `__proto__` property (as for all  other objects).

So if we set the `prototype` of a constructor, we can ensure that all objects created with that constructor are given that prototype:

```js
const personPrototype = {
  greet() {
    console.log(`hello, my name is ${this.name}!`);
  }
}

function Person(name) {
  this.name = name;
}

Person.prototype = personPrototype;
Person.prototype.constructor = Person;
```

Here we create:
* an object `personPrototype`, which has a `greet()` method
* a `Person()` constructor function which initializes the name of the person to create.

We then set the `Person` function's `prototype` property to point to `personPrototype`.

> The last line (`Person.prototype.constructor = Person;`) is needed because the prototype has a `constructor` property that points to the constructor function that was used to create the object. Since we have replaced the `prototype` property, we need to set `prototype.constructor` back to its original value.

Reviewer: Following on from https://github.com/mdn/content/pull/11373#discussion_r777624481 - this is better but still confusing. Perhaps we can make this more clear by just state what we're achieving:

```suggestion
The last line (`Person.prototype.constructor = Person;`) sets the prototype constructor property to the function used to create `Person` objects.
This is required because after setting `Person.prototype = personPrototype;` the property points to the constructor for the `personPrototype`, which is `Object` rather than `Person` (because `personPrototype` was constructed as an object literal).
```

Thanks for the explanation. I see why you want this to be an aside-generally you don't directly call this or need to call it. I did think of adding a note to that effect, but decided that at this point "just follow the rule" is better.
